src/graph.rs

```diff
@@ -396,90 +401,116 @@ pub(crate) fn parse_module(
   content: &str,
   maybe_resolver: &Option<Box<dyn Resolver>>,
 ) -> ModuleSlot {
-  // Init the module and determine its media type
-  let mut module = Module::new(specifier.clone(), content.to_string());
-  module.media_type = if let Some(headers) = maybe_headers {
-    if let Some(content_type) = headers.get("content-type") {
-      MediaType::from_content_type(&module.specifier, content_type)
-    } else {
-      MediaType::from(&module.specifier)
-    }
-  } else {
-    MediaType::from(&module.specifier)
-  };
-
   // Parse the module and start analyzing the module.
-  match ast::parse(&module.specifier, &module.source, module.media_type) {
+  match ast::parse(specifier, content, get_media_type(specifier, maybe_headers))
+  {
     Ok(parsed_module) => {
-      // Analyze the TypeScript triple-slash references
-      for reference in parsed_module.analyze_ts_references() {
-        match reference {
-          ast::TypeScriptReference::Path(specifier, range) => {
-            let resolved_dependency =
-              resolve(&specifier, &module.specifier, &range, maybe_resolver);
-            let dep = module.dependencies.entry(specifier).or_default();
-            dep.maybe_code = resolved_dependency;
-          }
-          ast::TypeScriptReference::Types(specifier, range) => {
-            let resolved_dependency =
-              resolve(&specifier, &module.specifier, &range, maybe_resolver);
-            if module.media_type == MediaType::JavaScript
-              || module.media_type == MediaType::Jsx
-            {
-              module.maybe_types_dependency =
-                Some((specifier, resolved_dependency));
-            } else {
-              let dep = module.dependencies.entry(specifier).or_default();
-              dep.maybe_type = resolved_dependency;
-            }
-          }
-        }
-      }
+      // Return the module as a valid module
+      ModuleSlot::Module(parse_module_from_ast(
+        specifier,
+        maybe_headers,
+        content,
+        &parsed_module,
+        maybe_resolver,
+      ))
+    }
+    Err(diagnostic) => ModuleSlot::Err(diagnostic.into()),
+  }
+}
 
-      // Analyze the X-TypeScript-Types header
-      if module.maybe_types_dependency.is_none() {
-        if let Some(headers) = maybe_headers {
-          if let Some(types_header) = headers.get("x-typescript-types") {
-            let resolved_dependency = resolve(
-              types_header,
-              &module.specifier,
-              &ast::Range::default(),
-              maybe_resolver,
-            );
-            module.maybe_types_dependency =
-              Some((types_header.clone(), resolved_dependency));
-          }
+pub(crate) fn parse_module_from_ast<
+  TComments: swc_common::comments::Comments,
+>(
+  specifier: &ModuleSpecifier,
+  maybe_headers: &Option<HashMap<String, String>>,
+  content: &str,
+  parsed_module: &impl ParsedModule<TComments>,
```

Comment:
Are these somewhat redundant?  I mean why does the `ParsedModule` trait provide a reference to the content, since it needed it _before_ it could have ever been generated?

Reply:
Forgot to address this. Will look tomorrow.

Reply:
Actually, I just looked into it. So we don't store the content in `ParsedAst` at the moment. It parses from a reference to the string. We also don't force people to implement a trait of `ParsedAst` that needs to provide the content. I think that's probably better because then we save a little memory and don't force the source file text to live in two places.

Reply:
Ok, I went through and made it so that `ParsedAst` has a `content(&self) -> Arc<String>` and this `Arc<String>` is shared with the `Module`. Let me know your thoughts as I'm not sure what's best here.